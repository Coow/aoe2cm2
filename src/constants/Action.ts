enum Action {
    PICK = 'PICK',
    BAN = 'BAN',
    SNIPE = 'SNIPE',
    STEAL = 'STEAL',
    REVEAL_ALL = 'REVEAL_ALL',
    REVEAL_PICKS = 'REVEAL_PICKS',
    REVEAL_BANS = 'REVEAL_BANS',
    REVEAL_SNIPES = 'REVEAL_SNIPES',
    PAUSE = 'PAUSE',
    RESET_CL = 'RESET_CL',
}

export default Action;