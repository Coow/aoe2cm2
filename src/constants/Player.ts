enum Player {
    HOST = 'HOST',
    GUEST = 'GUEST',
    NONE = 'NONE',
    SPEC = 'SPEC',
}

export default Player;