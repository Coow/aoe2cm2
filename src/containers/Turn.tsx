import {ApplicationState} from '../types';
import {connect} from 'react-redux';
import * as actions from "../actions";
import Turn from "../components/draft/Turn";
import {Dispatch} from 'redux';


export function mapStateToProps(state: ApplicationState) {
    return {
        nextAction: state.ownProperties.nextAction,
        highlightedAction: state.ownProperties.highlightedAction,
    };
}

export function mapDispatchToProps(dispatch: Dispatch<actions.Action>) {
    return {
        onHighlightedActionChanged: (value: number | null) => dispatch(actions.setHighlightedAction(value)),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(Turn);