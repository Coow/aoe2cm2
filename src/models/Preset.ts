import Civilisation from "./Civilisation";
import Turn from "./Turn";
import {CivilisationEncoder} from "../util/CivilisationEncoder";
import {Assert} from "../util/Assert";

class Preset {

    public static readonly EMPTY: Preset = new Preset('', [], []);

    public static readonly NEW: Preset = new Preset('', Civilisation.ALL, []);

    public static readonly SAMPLE: Preset = new Preset('Default Preset', Civilisation.ALL, [
        Turn.HOST_GLOBAL_BAN,
        Turn.GUEST_GLOBAL_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_GLOBAL_PICK,
        Turn.GUEST_GLOBAL_PICK,
        Turn.HOST_PICK,
        Turn.GUEST_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_SNIPE,
        Turn.GUEST_HIDDEN_SNIPE,
        Turn.REVEAL_ALL
    ]);

    public static readonly SIMPLE: Preset = new Preset('Simple Preset', Civilisation.ALL, [
        Turn.HOST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_PICK,
        Turn.HOST_NONEXCLUSIVE_PICK
    ]);
    public readonly name: string;
    public presetId?: string
    public readonly encodedCivilisations: string;
    public readonly turns: Turn[];

    constructor(name: string, civilisations: Civilisation[], turns: Turn[] = [], presetId?: string) {
        this.name = name;
        this.presetId = presetId;
        this.encodedCivilisations = CivilisationEncoder.encodeCivilisationArray(civilisations);
        this.turns = turns;
    }

    public static fromPojo(preset: { name: string, encodedCivilisations: string, turns: Turn[], presetId?: string } | undefined): Preset | undefined {
        if (preset === undefined) {
            return undefined;
        }
        Assert.isString(preset.name);
        Assert.isString(preset.encodedCivilisations);
        Assert.isOptionalString(preset.presetId);
        return new Preset(preset.name, CivilisationEncoder.decodeCivilisationArray(preset.encodedCivilisations), Turn.fromPojoArray(preset.turns), preset.presetId);
    }

    public addTurn(turn: Turn) {
        this.turns.push(turn);
    }

    get civilisations(): Civilisation[] {
        return CivilisationEncoder.decodeCivilisationArray(this.encodedCivilisations);
    }
}

export default Preset;