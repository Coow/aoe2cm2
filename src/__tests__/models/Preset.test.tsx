import Preset from "../../models/Preset";
import Turn from "../../models/Turn";
import Exclusivity from "../../constants/Exclusivity";
import Player from "../../constants/Player";
import Action from "../../constants/Action";

it('preset from invalid pojo throws', () => {
    expect(() => {
        Preset.fromPojo({turns: [{} as Turn, {} as Turn]} as { name: string, encodedCivilisations: string, turns: Turn[] });
    }).toThrowError("Expected argument to be string, but was undefined");
});

it('preset from valid pojo works', () => {
    const fromPojo = Preset.fromPojo({
        name: "some-name",
        encodedCivilisations: "0x1",
        turns: [{
            player: "GUEST",
            action: "PICK",
            exclusivity: Exclusivity.NONEXCLUSIVE,
            hidden: false,
            parallel: false
        }]
    } as { name: string, encodedCivilisations: string, turns: Turn[] });
    expect(fromPojo).toMatchSnapshot();
});

it('old turns without executingPlayer properties can be deserialised', () => {
    const pojo = {
        name: "Preset name",
        encodedCivilisations: "0x1",
        turns: [{
            id: "mocked-uuid",
            player: Player.HOST,
            action: Action.PICK,
            exclusivity: Exclusivity.NONEXCLUSIVE,
            hidden: false,
            parallel: false
        } as Turn]
    };
    const preset = Preset.fromPojo(pojo) as Preset;
    expect(preset.turns[0].executingPlayer).toEqual(Player.HOST);
});


it('new turns with executingPlayer properties can be deserialised', () => {
    const pojo = {
        name: "Preset name",
        encodedCivilisations: "0x1",
        turns: [{
            id: "mocked-uuid",
            player: Player.HOST,
            action: Action.PICK,
            exclusivity: Exclusivity.NONEXCLUSIVE,
            hidden: false,
            parallel: false,
            executingPlayer: Player.GUEST
        } as Turn]
    };
    const preset = Preset.fromPojo(pojo) as Preset;
    expect(preset.turns[0].executingPlayer).toEqual(Player.GUEST);
});

