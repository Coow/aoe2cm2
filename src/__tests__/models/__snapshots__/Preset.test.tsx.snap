// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preset from valid pojo works 1`] = `
Preset {
  "encodedCivilisations": "1",
  "name": "some-name",
  "presetId": undefined,
  "turns": Array [
    Turn {
      "action": "PICK",
      "exclusivity": "NONEXCLUSIVE",
      "executingPlayer": "GUEST",
      "hidden": false,
      "id": "mocked-uuid",
      "parallel": false,
      "player": "GUEST",
    },
  ],
}
`;
